File: runtime/python/prompty/prompty/harness/adapters.py
"""Implement dependency-free reference adapters for harness protocols."""

from __future__ import annotations

import inspect
import json
from collections.abc import Awaitable, Callable
from pathlib import Path
from time import perf_counter
from typing import Any

from ..model import (
    Checkpoint,
    HostToolRequest,
    HostToolResult,
    PermissionDecision,
    PermissionRequest,
    SessionEvent,
    SessionSummary,
    TurnEvent,
)

ToolHandler = Callable[[dict[str, Any], HostToolRequest], Any | Awaitable[Any]]


def _checkpoint_key(session_id: str, checkpoint_id: str) -> tuple[str, str]:
    return (session_id, checkpoint_id)


def _require_checkpoint_key(checkpoint: Checkpoint) -> tuple[str, str]:
    if not checkpoint.session_id:
        raise ValueError("Checkpoint session_id is required")
    if not checkpoint.id:
        raise ValueError("Checkpoint id is required")
    return _checkpoint_key(checkpoint.session_id, checkpoint.id)


def _error_message(error: BaseException) -> str:
    return str(error)


class CollectingEventSink:
    """Capture emitted turn and session events in memory."""

    def __init__(self) -> None:
        self.turn_events: list[TurnEvent] = []
        self.session_events: list[SessionEvent] = []

    def emit_turn(self, turn_event: TurnEvent) -> bool:
        """Append a turn event."""
        self.turn_events.append(turn_event)
        return True

    def emit_session(self, session_event: SessionEvent) -> bool:
        """Append a session event."""
        self.session_events.append(session_event)
        return True


class JsonlTraceWriter:
    """Append replayable trace records as newline-delimited JSON."""

    def __init__(self, path: str | Path) -> None:
        self.path = Path(path)
        self.path.parent.mkdir(parents=True, exist_ok=True)
        self._closed = False

    def append_turn(self, turn_event: TurnEvent) -> bool:
        """Append a turn event record."""
        return self._write({"kind": "turn", "event": turn_event.save()})

    def append_session(self, session_event: SessionEvent) -> bool:
        """Append a session event record."""
        return self._write({"kind": "session", "event": session_event.save()})

    def close(self, summary: SessionSummary | None) -> bool:
        """Append an optional summary and close the writer."""
        if summary is not None:
            if not self._write({"kind": "summary", "summary": summary.save()}):
                return False
        self._closed = True
        return True

    def _write(self, record: dict[str, Any]) -> bool:
        if self._closed:
            return False
        with self.path.open("a", encoding="utf-8") as file:
            file.write(json.dumps(record, separators=(",", ":")))
            file.write("\n")
        return True


class InMemoryCheckpointStore:
    """Store checkpoints in memory by session and checkpoint identifier."""

    def __init__(self) -> None:
        self._checkpoints: dict[tuple[str, str], Checkpoint] = {}

    async def save(self, checkpoint: Checkpoint) -> Checkpoint:
        """Persist a checkpoint in memory."""
        self._checkpoints[_require_checkpoint_key(checkpoint)] = checkpoint
        return checkpoint

    async def load(self, session_id: str, checkpoint_id: str) -> Checkpoint | None:
        """Load a checkpoint by identifiers."""
        return self._checkpoints.get(_checkpoint_key(session_id, checkpoint_id))

    async def list_checkpoints(self, session_id: str) -> list[Checkpoint]:
        """List checkpoints for a session."""
        return [checkpoint for checkpoint in self._checkpoints.values() if checkpoint.session_id == session_id]


class AllowAllPermissionResolver:
    """Resolve every permission request as approved."""

    async def request(self, request: PermissionRequest) -> PermissionDecision:
        """Return an approved permission decision."""
        return PermissionDecision(
            request_id=request.request_id,
            tool_call_id=request.tool_call_id,
            permission=request.permission,
            approved=True,
            reason="allow_all",
        )


class DenyAllPermissionResolver:
    """Resolve every permission request as denied."""

    async def request(self, request: PermissionRequest) -> PermissionDecision:
        """Return a denied permission decision."""
        return PermissionDecision(
            request_id=request.request_id,
            tool_call_id=request.tool_call_id,
            permission=request.permission,
            approved=False,
            reason="deny_all",
        )


class FunctionHostToolExecutor:
    """Dispatch host tool requests to registered local callables."""

    def __init__(self, handlers: dict[str, ToolHandler]) -> None:
        self.handlers = handlers

    async def execute(self, request: HostToolRequest) -> HostToolResult:
        """Execute a registered host tool callable."""
        started = perf_counter()
        handler = self.handlers.get(request.tool_name)
        if handler is None:
            return HostToolResult(
                request_id=request.request_id,
                tool_call_id=request.tool_call_id,
                tool_name=request.tool_name,
                success=False,
                error_kind="not_found",
                result={"message": f"No host tool registered for '{request.tool_name}'"},
                duration_ms=(perf_counter() - started) * 1000,
            )

        try:
            result = handler(request.arguments or {}, request)
            if inspect.isawaitable(result):
                result = await result
            return HostToolResult(
                request_id=request.request_id,
                tool_call_id=request.tool_call_id,
                tool_name=request.tool_name,
                success=True,
                result=result,
                duration_ms=(perf_counter() - started) * 1000,
            )
        except Exception as error:
            return HostToolResult(
                request_id=request.request_id,
                tool_call_id=request.tool_call_id,
                tool_name=request.tool_name,
                success=False,
                error_kind="exception",
                result={"message": _error_message(error)},
                duration_ms=(perf_counter() - started) * 1000,
            )

File: runtime/typescript/packages/core/src/harness/adapters.ts
// Copyright (c) Microsoft. All rights reserved.

import { appendFileSync, mkdirSync } from "node:fs";
import { dirname } from "node:path";

import {
  Checkpoint,
  HostToolRequest,
  HostToolResult,
  PermissionDecision,
  PermissionRequest,
  SessionEvent,
  SessionSummary,
  TurnEvent,
  type CheckpointStore,
  type EventSink,
  type HostToolExecutor,
  type PermissionResolver,
  type TraceWriter,
} from "../model/index.js";

type JsonRecord = Record<string, unknown>;
type ToolHandler = (args: JsonRecord, request: HostToolRequest) => unknown | Promise<unknown>;

function checkpointKey(sessionId: string, checkpointId: string): string {
  return `${sessionId}\u0000${checkpointId}`;
}

function requireCheckpointKey(checkpoint: Checkpoint): { sessionId: string; checkpointId: string } {
  if (!checkpoint.sessionId) {
    throw new Error("Checkpoint sessionId is required");
  }
  if (!checkpoint.id) {
    throw new Error("Checkpoint id is required");
  }
  return { sessionId: checkpoint.sessionId, checkpointId: checkpoint.id };
}

function errorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

/** Captures emitted turn and session events in memory. */
export class CollectingEventSink implements EventSink {
  readonly turnEvents: TurnEvent[] = [];
  readonly sessionEvents: SessionEvent[] = [];

  emitTurn(turnEvent: TurnEvent): boolean {
    this.turnEvents.push(turnEvent);
    return true;
  }

  emitSession(sessionEvent: SessionEvent): boolean {
    this.sessionEvents.push(sessionEvent);
    return true;
  }
}

/** Appends replayable trace records as newline-delimited JSON. */
export class JsonlTraceWriter implements TraceWriter {
  private closed = false;

  constructor(private readonly path: string) {
    mkdirSync(dirname(path), { recursive: true });
  }

  appendTurn(turnEvent: TurnEvent): boolean {
    return this.write({ kind: "turn", event: turnEvent.save() });
  }

  appendSession(sessionEvent: SessionEvent): boolean {
    return this.write({ kind: "session", event: sessionEvent.save() });
  }

  close(summary: SessionSummary | null): boolean {
    if (summary) {
      if (!this.write({ kind: "summary", summary: summary.save() })) {
        return false;
      }
    }
    this.closed = true;
    return true;
  }

  private write(record: JsonRecord): boolean {
    if (this.closed) {
      return false;
    }
    appendFileSync(this.path, `${JSON.stringify(record)}\n`, "utf8");
    return true;
  }
}

/** Stores checkpoints in memory by session and checkpoint identifier. */
export class InMemoryCheckpointStore implements CheckpointStore {
  private readonly checkpoints = new Map<string, Checkpoint>();

  async save(checkpoint: Checkpoint): Promise<Checkpoint> {
    const { sessionId, checkpointId } = requireCheckpointKey(checkpoint);
    this.checkpoints.set(checkpointKey(sessionId, checkpointId), checkpoint);
    return checkpoint;
  }

  async load(sessionId: string, checkpointId: string): Promise<Checkpoint | null> {
    return this.checkpoints.get(checkpointKey(sessionId, checkpointId)) ?? null;
  }

  async listCheckpoints(sessionId: string): Promise<Checkpoint[]> {
    return [...this.checkpoints.values()].filter((checkpoint) => checkpoint.sessionId === sessionId);
  }
}

/** Resolves every permission request as approved. */
export class AllowAllPermissionResolver implements PermissionResolver {
  async request(request: PermissionRequest): Promise<PermissionDecision> {
    return new PermissionDecision({
      requestId: request.requestId,
      toolCallId: request.toolCallId,
      permission: request.permission,
      approved: true,
      reason: "allow_all",
    });
  }
}

/** Resolves every permission request as denied. */
export class DenyAllPermissionResolver implements PermissionResolver {
  async request(request: PermissionRequest): Promise<PermissionDecision> {
    return new PermissionDecision({
      requestId: request.requestId,
      toolCallId: request.toolCallId,
      permission: request.permission,
      approved: false,
      reason: "deny_all",
    });
  }
}

/** Dispatches host tool requests to registered local functions. */
export class FunctionHostToolExecutor implements HostToolExecutor {
  constructor(private readonly handlers: Record<string, ToolHandler>) {}

  async execute(request: HostToolRequest): Promise<HostToolResult> {
    const started = Date.now();
    const handler = this.handlers[request.toolName];
    if (!handler) {
      return new HostToolResult({
        requestId: request.requestId,
        toolCallId: request.toolCallId,
        toolName: request.toolName,
        success: false,
        errorKind: "not_found",
        result: { message: `No host tool registered for '${request.toolName}'` },
        durationMs: Date.now() - started,
      });
    }

    try {
      const result = await handler(request.arguments ?? {}, request);
      return new HostToolResult({
        requestId: request.requestId,
        toolCallId: request.toolCallId,
        toolName: request.toolName,
        success: true,
        result,
        durationMs: Date.now() - started,
      });
    } catch (error) {
      return new HostToolResult({
        requestId: request.requestId,
        toolCallId: request.toolCallId,
        toolName: request.toolName,
        success: false,
        errorKind: "exception",
        result: { message: errorMessage(error) },
        durationMs: Date.now() - started,
      });
    }
  }
}
